### PR TITLE
feat(Request): Add support for several proxy "forwarded" headers

### PR DESCRIPTION
**Pull request by kgriffs**
Add support for some "forwarded" headers, included several new
attributes and a reworking of some of the existing code to better
facilitate code reuse and performance.

Also clean up a couple nits in the docstrings for the sake of
consistency.

### CODE REVIEW

File: falcon/media/base.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 class BaseHandler(object):
     """Abstract Base Class for an internet media type handler"""
 
-    @abc.abstractmethod
+    @abc.abstractmethod  # pragma: no cover
```
Comment (kgriffs) [inline falcon/media/base.py]: Coverage was giving me grief on this line and the one below for no good reason that I could tell.

File: falcon/request.py [hunk 2]
```diff
@@ -1495,3 +1697,36 @@ def __init__(self):
...
+    """
+
+    # NOTE(kgriffs): Use "client" since "for" is a keyword, and
+    # "scheme" instead of "proto" to be consistent with the
+    # falcon.Request interface.
+    __slots__ = ('by', 'client', 'host', 'scheme')
+
+    def __init__(self):
+        self.by = None
+        self.client = None
```
Comment (kgriffs) [inline falcon/request.py]: If you can think of a better name for this attr, please let me know. Per the inline comment, using "for" results in a SyntaxError on reference, and "forwarded_for" seemed a bit unwieldy.

Comment (jmvrbanac) [inline falcon/request.py]: Maybe target or destination?

Comment (kgriffs) [inline falcon/request.py]: TBH, I think target or destination more accurately describe the "by" parameter. According to RFC 7239:

> The "for" parameter is used to disclose information about the client that initiated the request and subsequent proxies in a chain of proxies.

Whereas:

>  The "by" parameter is used to disclose the interface where the request came in to the proxy server.

What do you think about the following variations?

```py
forwarded = req.forwarded[0]

forwarded.source
forwarded.dest

forwarded.source
forwarded.dest

forwarded.client
forwarded.target

forwarded.client
forwarded.dest

forwarded.client
forwarded.proxy

forwarded.client
forwarded.by
```

Comment (jmvrbanac) [inline falcon/request.py]: I think "destination" seems appropriate and generic enough
```
forwarded.source
forwarded.destination
--- or ---
forwarded.dest
```

Comment (kgriffs) [inline falcon/request.py]: Sounds good, I'll make the change. Thanks!

**Comment by kgriffs** [comment]
OK, I've updated this now with the new names. Let me know how it looks and if I missed anything!

**Pull request merged**